crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2059,6 +2114,293 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 }
 
+/// Annotation expressions.
+impl<'db> TypeInferenceBuilder<'db> {
+    fn annotation_expression(&mut self, expression: &ast::Expr) -> Type<'db> {
+        let _span = tracing::trace_span!(
+            "infer_annotation_expression",
+            expr = ?expression
+        )
+        .entered();
+        // https://typing.readthedocs.io/en/latest/spec/annotations.html#grammar-token-expression-grammar-annotation_expression
+        let ty = match expression {
+            // Forms which are possibly valid type expressions.
+            type_expr @ (ast::Expr::NoneLiteral(..) | ast::Expr::Name(..)) => {
+                self.type_expression(type_expr)
+            }
+
+            // NOTE: This section is presently identical with the corresponding section in the type
+            // expressions implementation, but they may diverge from each other in the future,
+            // especially after the introduction of special forms.
```

Comment:
I think I would prefer a different approach here, to reduce duplication. I think `infer_annotation_expression` should handle any valid annotation-expression-only forms, and just defer everything else to `infer_type_expression` blindly. Because any valid type expression is also a valid annotation expression, and we shouldn't need to double-encode information here in `infer_annotation_expression` about what is or isn't a valid type expression.

Given that as of this PR we aren't handling _any_ annotation-expression-only forms, that means that for now, `infer_annotation_expression` should be just a comment like `TODO handle annotation-expression-only special forms` and then  just defer everything to `infer_type_expression`.

Are there problems with this approach that I'm missing?

Reply:
That approach seems perfectly reasonable. I debated doing it that way initially, and simply didn’t have a strong enough sense of the tradeoff there—reviewing the grammar, though, I see what you mean. There just isn’t that much which is specific to annotation expressions.

Reply:
Done (see 6af652ce)!